main.go

```diff
@@ -105,72 +106,125 @@ func mkServiceLogs(logConf logconfig.LoggingConfig) (syncLog, clusterSyncLog, ac
+			if err != nil {
+				log.Printf("Handler initialization failure %s", err)
+			}
+			s.handler = handler
+			log.Println("Handler replaced")
+		case <-intr:
+			log.Println("Shutting down")
+			err := s.srv.Shutdown(s.ctx)
+			if err != nil {
+				log.Printf("Server shutsown error: %s", err)
+			}
+			log.Println("Fin")
+		}
+	}
+}
+
+func (s *service) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	handler := s.handler
+	handler.ServeHTTP(rw, r)
+}
+func (s *service) createHandler(conf config.Config) (http.Handler, error) {
+	transportMatcher, err := transport.ConfigureHTTPTransports(conf.Service.Client)
 	if err != nil {
-		log.Fatalf("Couldn't set up client Transports - err: %q", err)
+		return nil, fmt.Errorf("Couldn't set up client Transports - err: %q", err)
 	}
-	syncLog, clusterSyncLog, accessLog, err := mkServiceLogs(s.config.Logging)
+	syncLog, clusterSyncLog, accessLog, err := mkServiceLogs(conf.Logging)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	methods := make(map[string]struct{})
-	for _, method := range s.config.Logging.SyncLogMethods {
+	for _, method := range conf.Logging.SyncLogMethods {
 		methods[method] = struct{}{}
 	}
 
-	crdstore.InitializeCredentialsStore(s.config.CredentialsStore)
+	crdstore.InitializeCredentialsStore(conf.CredentialsStore)
 	syncSender := &storages.SyncSender{SyncLog: syncLog, AllowedMethods: methods}
 	storage, err := storages.InitStorages(
 		transportMatcher,
-		s.config.Clusters,
-		s.config.Backends,
+		conf.Clusters,
+		conf.Backends,
 		syncSender)
 
 	if err != nil {
 		log.Fatalf("Storages initialization problem: %q", err)
+		return nil, err
 	}
 
-	regionsRT, err := regions.NewRegions(s.config.Regions, storage, clusterSyncLog)
+	regionsRT, err := regions.NewRegions(conf.Regions, storage, clusterSyncLog)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	regionsDecoratedRT := httphandler.DecorateRoundTripper(s.config.Service.Client,
-		accessLog, s.config.Service.Server.HealthCheckEndpoint, regionsRT)
-
-	handler, err := httphandler.NewHandlerWithRoundTripper(regionsDecoratedRT, s.config.Service.Server)
-	if err != nil {
-		return err
-	}
+	regionsDecoratedRT := httphandler.DecorateRoundTripper(conf.Service.Client,
+		accessLog, conf.Service.Server.HealthCheckEndpoint, regionsRT)
 
-	err = metrics.Init(s.config.Metrics)
+	handler, err := httphandler.NewHandlerWithRoundTripper(regionsDecoratedRT, conf.Service.Server)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	srv := &graceful.Server{
-		Server: &http.Server{
-			Addr:         s.config.Service.Server.Listen,
-			Handler:      handler,
-			ReadTimeout:  s.config.Service.Server.ReadTimeout.Duration,
-			WriteTimeout: s.config.Service.Server.WriteTimeout.Duration,
-		},
-		Timeout: s.config.Service.Server.ShutdownTimeout.Duration,
-	}
-
-	srv.SetKeepAlivesEnabled(true)
-	listener, err := net.Listen("tcp", s.config.Service.Server.Listen)
+	err = metrics.Init(conf.Metrics)
```

Comment:
My proposition it's moving `err = metrics.Init(conf.Metrics)` line outside HUP signal block - I suspect many goroutines on stack after recive many HUP signals.